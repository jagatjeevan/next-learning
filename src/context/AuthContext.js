import { createContext } from '../utils/create-context';

const auth = {
  username: '',
  token: null,
  profile: {},
};

const reducer = (state, action) => {
  switch (action.type) {
    case 'LOGOUT':
      return { username: '', token: null, profile: {} };
    case 'LOGIN':
      return {
        username: action.payload.name,
        token: action.payload.token,
        profile: action.payload.profile,
      };
    default:
      return { ...state };
  }
};

const logout = (dispatch) => () => dispatch({ type: 'LOGOUT' });
const login = (dispatch) => (loginObj) => dispatch({ type: 'LOGIN', payload: loginObj });

export const { Context, Provider } = createContext(reducer, { logout, login }, auth);
