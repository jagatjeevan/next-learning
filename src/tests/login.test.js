import { act, fireEvent, getByText, render, screen, waitFor } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Login from '../../pages/login';
import { Provider as AuthContextProvider } from '../Context/AuthContext';

describe('Login', () => {
  it('renders the login page when user is not logged in', () => {
    const authValue = {
      username: '',
      token: null,
      profile: {},
    };
    const login = jest.mock();
    const logout = jest.mock();
    const tree = renderer
      .create(
        <AuthContextProvider value={{ state: authValue, login, logout }}>
          <Login />
        </AuthContextProvider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders the login page when user is logged in', () => {
    const authValue = {
      username: 'Jeevan',
      token: 'skdjflksjdf',
      profile: {},
    };
    const login = jest.mock();
    const logout = jest.mock();
    const tree = renderer
      .create(
        <AuthContextProvider value={{ state: authValue, login, logout }}>
          <Login />
        </AuthContextProvider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('logs in the user on login button click', async () => {
    const authValue = {
      username: '',
      token: null,
      profile: {},
    };
    const login = jest.mock();
    const logout = jest.mock();
    render(
      <AuthContextProvider value={{ state: authValue, login, logout }}>
        <Login />
      </AuthContextProvider>,
    );
    const buttonComponent = screen.getByRole('button');
    expect(buttonComponent).toBeInTheDocument();
    fireEvent.click(buttonComponent);
    act(() => {
      expect(screen.getByText(/Welcome Jeevan/i)).toBeInTheDocument();
    });
  });
});
