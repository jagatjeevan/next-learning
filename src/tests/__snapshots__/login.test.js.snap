// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders the login page when user is logged in 1`] = `
Array [
  <h1>
    Login page
  </h1>,
  <a
    href="/"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    Home page
  </a>,
  <button
    onClick={[Function]}
    type="button"
  >
    Login
  </button>,
]
`;

exports[`Login renders the login page when user is not logged in 1`] = `
Array [
  <h1>
    Login page
  </h1>,
  <a
    href="/"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    Home page
  </a>,
  <button
    onClick={[Function]}
    type="button"
  >
    Login
  </button>,
]
`;
